### PR TITLE
Making BulkRequest Releasable, and updating all tests to release it

### PR DESCRIPTION
**Pull request by masseyke**
This is the first step toward allowing bulk requests to reuse an underlying byte array rather than copying off smaller byte arrays for each request that later have to be garbage collected. There will be many PRs. The basic outline will be:

1. Make BulkRequest RefCounted, making sure that its ref count is always zero when it is garbage collected and never zero when still in use.
2. Do the same for BulkShardRequest
3. Do the same for the individual requests used by BulkRequest and BulkShardRequest -- BulkItemRequest, IndexRequest, UpdateRequest, and ReindexRequest.
4. Update IndexRequest to use the underlying shared byte array for its source

This PR is the first part of that first item. In preparation for making BulkRequest implement RefCounted, I am updating it to implement Releasable. Then I am updating all tests to properly close it. In a subsequent PR I will have BulkRequest implement RefCounted and update the non-test code to properly use it. Since there are well over 100 tests that need to be updated, I'm handling those first so that they don't clutter the next PR.

### CODE REVIEW

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/TSDBIndexingIT.java [hunk 1]
```diff
@@ -456,16 +456,17 @@ public void testTrimId() throws Exception {
         {
             Instant time = Instant.now();
             for (int i = 0; i < numBulkRequests; i++) {
-                BulkRequest bulkRequest = new BulkRequest(dataStreamName);
-                for (int j = 0; j < numDocsPerBulk; j++) {
-                    var indexRequest = new IndexRequest(dataStreamName).opType(DocWriteRequest.OpType.CREATE);
-                    indexRequest.source(DOC.replace("$time", formatInstant(time)), XContentType.JSON);
-                    bulkRequest.add(indexRequest);
-                    time = time.plusMillis(1);
+                try (BulkRequest bulkRequest = new BulkRequest()) {
```
Comment (joegallo) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/TSDBIndexingIT.java:459]: `new BulkRequest(dataStreamName)` became `new BulkRequest()` -- does it matter?

File: server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java [hunk 2]
```diff
@@ -79,6 +79,7 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
             ThreadContext threadContext = threadPool.getThreadContext();
             assertEquals(headerValue, threadContext.getHeader(headerKey));
             assertSame(transientValue, threadContext.getTransient(transientKey));
+            listener.onResponse(new BulkResponse(new BulkItemResponse[0], 1L));
```
Comment (joegallo) [inline server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:82]: I'm curious why this becomes necessary (I'm not pushing back on it, I'm just interested).

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java [hunk 3]
```diff
@@ -442,7 +442,7 @@ public void testAuthnWithInvalidatedToken() throws Exception {
         TokenService tokenService = createTokenService(tokenServiceEnabledSettings, systemUTC());
         Authentication authentication = AuthenticationTestHelper.builder()
             .user(new User("joe", "admin"))
-            .realmRef(new RealmRef("native_realm", "native", "node1"))
+            .realmRef(new RealmRef("native_realm", "na7tive", "node1"))
```
Comment (joegallo) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:445]: ?

Comment (masseyke) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java]: Oops.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/template/SimpleIndexTemplateIT.java [hunk 4]
```diff
@@ -715,13 +716,15 @@ public void testStrictAliasParsingInIndicesCreatedViaTemplates() throws Exceptio
...
             .get();
 
-        prepareIndex("a1").setId("test").setSource("{}", XContentType.JSON).get();
-        BulkResponse response = client().prepareBulk().add(new IndexRequest("a2").id("test").source("{}", XContentType.JSON)).get();
-        assertThat(response.hasFailures(), is(false));
-        assertThat(response.getItems()[0].isFailed(), equalTo(false));
-        assertThat(response.getItems()[0].getIndex(), equalTo("a2"));
-        assertThat(response.getItems()[0].getId(), equalTo("test"));
-        assertThat(response.getItems()[0].getVersion(), equalTo(1L));
+        client().prepareIndex("a1").setId("test").setSource("{}", XContentType.JSON).get();
```
Comment (joegallo) [inline server/src/internalClusterTest/java/org/elasticsearch/indices/template/SimpleIndexTemplateIT.java:719]: Does adding `client().` here change much? I haven't noticed a similar change elsewhere yet.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/TSDBIndexingIT.java (hunk 1 above)
Comment (masseyke) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/TSDBIndexingIT.java:459]: It probably doesn't, but I definitely didn't mean to remove it. Adding it back.

File: server/src/internalClusterTest/java/org/elasticsearch/indices/template/SimpleIndexTemplateIT.java (hunk 4 above)
Comment (masseyke) [reply server/src/internalClusterTest/java/org/elasticsearch/indices/template/SimpleIndexTemplateIT.java:719]: This is probably the result of a bad merge by me. At some point recently someone removed all of these client() calls. I'll remove it.

File: server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java (hunk 2 above)
Comment (masseyke) [reply server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:82]: In this PR I'm not actually asserting that the BulkRequest objects are closed (that comes in the next PR). But without this line the listener is never called so the BulkRequest is never called. Here's the stack where it closes the BulkRequest through this line (in a slightly different branch of code that will be in a future PR)

```
	at org.elasticsearch.action.bulk.BulkRequest.close(BulkRequest.java:503)
	at org.elasticsearch.core.IOUtils.close(IOUtils.java:71)
	at org.elasticsearch.core.Releasables.close(Releasables.java:34)
	at org.elasticsearch.core.Releasables.closeExpectNoException(Releasables.java:58)
	at org.elasticsearch.action.ActionListenerImplementations$2.run(ActionListenerImplementations.java:49)
	at org.elasticsearch.action.ActionListenerImplementations$RunAfterActionListener.onResponse(ActionListenerImplementations.java:262)
	at org.elasticsearch.action.ActionListener$3.onResponse(ActionListener.java:312)
	at org.elasticsearch.action.bulk.Retry$RetryHandler.finishHim(Retry.java:191)
	at org.elasticsearch.action.bulk.Retry$RetryHandler.onResponse(Retry.java:108)
	at org.elasticsearch.action.bulk.Retry$RetryHandler.onResponse(Retry.java:74)
	at org.elasticsearch.action.bulk.BulkProcessorTests.lambda$testBulkProcessorFlushPreservesContext$0(BulkProcessorTests.java:82)
	at org.elasticsearch.action.bulk.Retry$RetryHandler.execute(Retry.java:234)
	at org.elasticsearch.action.bulk.Retry$RetryHandler.execute(Retry.java:223)
	at org.elasticsearch.action.bulk.Retry.withBackoff(Retry.java:54)
	at org.elasticsearch.action.bulk.BulkRequestHandler.execute(BulkRequestHandler.java:59)
	at org.elasticsearch.action.bulk.BulkProcessor.execute(BulkProcessor.java:486)
	at org.elasticsearch.action.bulk.BulkProcessor.execute(BulkProcessor.java:496)
	at org.elasticsearch.action.bulk.BulkProcessor$Flush.run(BulkProcessor.java:537)
	at org.elasticsearch.threadpool.Scheduler$ReschedulingRunnable.doRun(Scheduler.java:223)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.threadpool.ThreadPool$1.run(ThreadPool.java:484)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by masseyke** [comment]
Closing in favor of multiple smaller PRs

**Pull request closed without merging**